File: app/src/test/java/com/coinninja/coinkeeper/cn/account/AddressCacheTest.java
package com.coinninja.coinkeeper.cn.account;

import com.coinninja.coinkeeper.cn.wallet.HDWallet;
import com.coinninja.coinkeeper.model.helpers.AddressHelper;
import com.coinninja.coinkeeper.model.helpers.WalletHelper;
import com.coinninja.coinkeeper.wallet.data.TestData;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.Arrays;

import static org.mockito.Matchers.anyInt;
import static org.mockito.Matchers.anyString;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class AddressCacheTest {
    private static final int numAddressesToCache = 10;

    @Mock
    private HDWallet hdWallet;

    @Mock
    private WalletHelper walletHelper;

    @Mock
    private AddressHelper addressHelper;

    private AddressCache addressCache;

    @Before
    public void setUp() {
        addressCache = new AddressCache(hdWallet, addressHelper, walletHelper, numAddressesToCache);
    }

    @After
    public void tearDown() {
        hdWallet = null;
        walletHelper = null;
        addressHelper = null;
        addressCache = null;
    }

    @Test
    public void requests_current_index_plus_padding() {
        int externalIndex = 22;
        int internalIndex = 2;
        when(walletHelper.getCurrentExternalIndex()).thenReturn(externalIndex);
        when(walletHelper.getCurrentInternalIndex()).thenReturn(internalIndex);
        when(hdWallet.fillBlock(anyInt(), anyInt(), anyInt())).thenReturn(new String[0]);

        addressCache.cacheAddressesFor(HDWallet.EXTERNAL);

        verify(hdWallet).fillBlock(HDWallet.EXTERNAL, 0, externalIndex + numAddressesToCache);

        addressCache.cacheAddressesFor(HDWallet.INTERNAL);

        verify(hdWallet).fillBlock(HDWallet.INTERNAL, 0, internalIndex + numAddressesToCache);
    }

    @Test
    public void caches_addresses_generated_by_hd_wallet() {
        int externalIndex = 5;
        when(walletHelper.getCurrentExternalIndex()).thenReturn(externalIndex);
        String[] addresses = Arrays.copyOfRange(TestData.INSTANCE.getEXTERNAL_ADDRESSES(), 0, 15);
        when(hdWallet.fillBlock(HDWallet.EXTERNAL, 0, 15)).thenReturn(addresses);

        addressCache.cacheAddressesFor(HDWallet.EXTERNAL);

        for (int i = 0; i < addresses.length; i++) {
            verify(addressHelper).saveAddress(HDWallet.EXTERNAL, i, addresses[i]);
        }

        verify(addressHelper, times(15)).saveAddress(eq(HDWallet.EXTERNAL), anyInt(), anyString());
    }

    @Test
    public void creates_internal_cache_only_when_necessary() {
        int internalIndex = 2;
        when(walletHelper.getCurrentInternalIndex()).thenReturn(internalIndex);
        when(addressHelper.getAddressCountFor(HDWallet.INTERNAL)).thenReturn(12);

        addressCache.cacheAddressesFor(HDWallet.INTERNAL);

        verify(hdWallet, times(0)).fillBlock(anyInt(), anyInt(), anyInt());
    }

    @Test
    public void creates_external_cache_only_when_necessary() {
        int externalIndex = 22;
        when(walletHelper.getCurrentExternalIndex()).thenReturn(externalIndex);
        when(addressHelper.getAddressCountFor(HDWallet.EXTERNAL)).thenReturn(32);

        addressCache.cacheAddressesFor(HDWallet.EXTERNAL);

        verify(hdWallet, times(0)).fillBlock(anyInt(), anyInt(), anyInt());
    }


}

File: app/src/test/java/com/coinninja/coinkeeper/cn/account/AddressCacheTest.kt
package com.coinninja.coinkeeper.cn.account

import app.coinninja.cn.libbitcoin.HDWallet
import app.coinninja.cn.libbitcoin.model.DerivationPath
import app.coinninja.cn.libbitcoin.model.MetaAddress
import com.coinninja.coinkeeper.model.db.Wallet
import com.coinninja.coinkeeper.wallet.data.TestData.EXTERNAL_ADDRESSES
import com.nhaarman.mockitokotlin2.*
import org.junit.Test

class AddressCacheTest {
    private fun createAddressCache(): AddressCache {
        val addressCache = AddressCache(mock(), mock(), mock(), 10)
        val wallet: Wallet = mock()
        whenever(wallet.purpose).thenReturn(49)
        whenever(wallet.coinType).thenReturn(0)
        whenever(wallet.accountIndex).thenReturn(0)
        whenever(addressCache.walletHelper.primaryWallet).thenReturn(wallet)
        return addressCache
    }

    @Test
    fun requests_current_index_plus_padding() {
        val addressCache = createAddressCache()
        val externalIndex = 22
        val internalIndex = 2
        val wallet = addressCache.walletHelper.primaryWallet
        whenever(wallet.externalIndex).thenReturn(externalIndex)
        whenever(wallet.internalIndex).thenReturn(internalIndex)
        whenever(addressCache.hdWallet.fillBlock(any(), any(), any(), any(), any(), any(), any())).thenReturn(emptyArray())

        addressCache.cacheAddressesFor(wallet, HDWallet.EXTERNAL)
        verify(addressCache.hdWallet).fillBlock(wallet, 49, 0, 0, HDWallet.EXTERNAL, 0, externalIndex + numAddressesToCache)
        addressCache.cacheAddressesFor(wallet, HDWallet.INTERNAL)
        verify(addressCache.hdWallet).fillBlock(wallet, 49, 0, 0, HDWallet.INTERNAL, 0, internalIndex + numAddressesToCache)
    }

    @Test
    fun caches_addresses_generated_by_hd_wallet() {
        val addressCache = createAddressCache()
        val externalIndex = 5
        val wallet = addressCache.walletHelper.primaryWallet
        whenever(wallet.externalIndex).thenReturn(externalIndex)

        val addresses: Array<MetaAddress> = createMockAddresses(0, 15)
        whenever(addressCache.hdWallet.fillBlock(wallet, 49, 0, 0, HDWallet.EXTERNAL, 0, 15)).thenReturn(addresses)
        addressCache.cacheAddressesFor(wallet, HDWallet.EXTERNAL)

        addresses.forEach {
            verify(addressCache.addressHelper).saveAddress(wallet, it)
        }

        verify(addressCache.addressHelper, times(15)).saveAddress(any(), any())
    }

    @Test
    fun creates_internal_cache_only_when_necessary() {
        val addressCache = createAddressCache()
        val internalIndex = 2
        val wallet: Wallet = mock()
        whenever(wallet.externalIndex).thenReturn(internalIndex)
        whenever(addressCache.addressHelper.getAddressCountFor(wallet, HDWallet.INTERNAL)).thenReturn(12)
        addressCache.cacheAddressesFor(wallet, HDWallet.INTERNAL)
        verify(addressCache.hdWallet, times(0)).fillBlock(any(), any(), any(), any(), any(), any(), any())
    }

    @Test
    fun creates_external_cache_only_when_necessary() {
        val addressCache = createAddressCache()
        val externalIndex = 22
        val wallet: Wallet = mock()
        whenever(wallet.externalIndex).thenReturn(externalIndex)
        whenever(addressCache.addressHelper.getAddressCountFor(wallet, HDWallet.EXTERNAL)).thenReturn(32)
        addressCache.cacheAddressesFor(wallet, HDWallet.EXTERNAL)
        verify(addressCache.hdWallet, times(0)).fillBlock(any(), any(), any(), any(), any(), any(), any())
    }

    private fun createMockAddresses(start: Int, end: Int): Array<MetaAddress> {
        val addresses: Array<String> = EXTERNAL_ADDRESSES.copyOfRange(start, end)
        val metaAddresses = mutableListOf<MetaAddress>()
        addresses.forEachIndexed { index, address ->
            metaAddresses.add(MetaAddress(address, "-- pubkey ${index} --", DerivationPath(49, 0, 0, 0, index)))
        }
        return metaAddresses.toTypedArray()
    }

    companion object {
        private const val numAddressesToCache = 10
    }
}